data_structures/bloom_filters/counting_bloom_filter/counting_bloom_filter.c

```diff
@@ -0,0 +1,234 @@
+        h ^= murmur_32_scramble(k);
+        h = (h << 13) | (h >> 19);
+        h = h * 5 + 0xe6546b64;
+    }
+    /* Read the rest. */
+    k = 0;
+    for (unsigned long i = len & 3; i; i--)
+    {
+        k <<= 8;
+        k |= key[i - 1];
+    }
+    // A swap is *not* necessary here because the preceding loop already
+    // places the low bytes in the low places according to whatever endianness
+    // we use. Swaps only apply when the memory is copied in a chunk.
+    h ^= murmur_32_scramble(k);
+    /* Finalize. */
+    h ^= len;
+    h ^= h >> 16;
+    h *= 0x85ebca6b;
+    h ^= h >> 13;
+    h *= 0xc2b2ae35;
+    h ^= h >> 16;
+    return h;
+}
+
+/**
+ * @brief  A union to store the counts as 4-bit buckets instead of 8.
+ *         Allows for the counting filter to be more space efficient.
+ */
+union Nibbler
+{
+    struct
+    {
+        unsigned int first : 4;
+        unsigned int second : 4;
+    } nibbles;
+    unsigned char byte_value;
+};
+
+struct CountingBloomFilter
+{
+    int hashFuncs;  ///< The number of hash functions the bloom filter will use.
+    int numSlots;   ///< The number of buckets in the underlying array.
+    int maxElems;   ///< The expected max number of elements to be inserted
+    union Nibbler buckets[];  ///< The underlying array structure of the filter
+};
+
+void countingBloom_printStats(struct CountingBloomFilter *this)
+{
+    int numOnes = 0;
+    for (int i = 0; i < this->numSlots; ++i)
+    {
+        float slot = ((float)i) / 2.0;
+        if (slot == (int)slot)
+        {
+            if (this->buckets[(int)slot].nibbles.first > 0)
+            {
+                ++numOnes;
+            }
+        }
+        else
+        {
+            if (this->buckets[(int)slot].nibbles.second > 0)
+            {
+                ++numOnes;
+            }
+        }
+    }
+
+    float nStar = -((float)this->numSlots / (float)this->hashFuncs) *
+                  log(1.0 - ((float)numOnes / (float)this->numSlots));
+
+    printf(
+        "Max Elements (n): %d \nHash Functions (k): %d \nSlots (m): "
+        "%d\nEstimated Elements Inserted (n*): %f (estimate deteriorates as "
+        "inserted elements approaches n)\n",
+        this->maxElems, this->hashFuncs, this->numSlots, nStar);
+}
+
+int countingBloom_insert(struct CountingBloomFilter *this, void *elem,
+                         size_t size)
+{
+    for (int i = 0; i < this->hashFuncs; ++i)
+    {
+        int spot = murmur3_32((uint8_t *)elem, size, i) % this->numSlots;
+        float slot = ((float)spot) / 2.0;
+        if (slot == (int)slot)
+        {
+            if (this->buckets[(int)slot].nibbles.first++ == 15)
+            {
+                return 1;
+            };
+        }
+        else
+        {
+            if (this->buckets[(int)slot].nibbles.second++ == 15)
+            {
+                return 1;
+            };
+        }
+    }
+    return 0;
+}
+int countingBloom_remove(struct CountingBloomFilter *this, void *elem,
```

Comment:
```suggestion
void countingBloom_remove(struct CountingBloomFilter *this, void *elem,
```

Reply:
remove returns an int to denote if the removal failed (a counter underflow would denote that an item that wasn't added is being removed). Can change this is needed but was intentional.

Reply:
> remove returns an int to denote if the removal failed (a counter underflow would denote that an item that wasn't added is being removed). Can change this is needed but was intentional.

Good point, guess I didn't read it carefully